SonarQube.TeamBuild.PreProcessor/Roslyn/EmbeddedAnalyzerInstaller.cs

```diff
@@ -0,0 +1,208 @@
+
+            if (Directory.Exists(cacheDir))
+            {
+                this.logger.LogDebug(RoslynResources.EAI_UsingCachedResource, cacheDir);
+            }
+            else
+            {
+                FetchResourceFromServer(plugin, cacheDir);
+            }
+
+            string[] allFiles = Directory.GetFiles(cacheDir, "*.*", SearchOption.AllDirectories);
+            return allFiles;
+        }
+
+        private void FetchResourceFromServer(Plugin plugin, string targetDir)
+        {
+            logger.LogDebug(RoslynResources.EAI_FetchingPluginResource, plugin.Key, plugin.Version, plugin.StaticResourceName);
+
+            Directory.CreateDirectory(targetDir);
+
+            if (server.TryDownloadEmbeddedFile(plugin.Key, plugin.StaticResourceName, targetDir))
+            {
+                string targetFilePath = Path.Combine(targetDir, plugin.StaticResourceName);
+
+                if (IsZipFile(targetFilePath))
+                {
+                    logger.LogDebug(Resources.MSG_ExtractingFiles, targetDir);
+                    ZipFile.ExtractToDirectory(targetFilePath, targetDir);
+                }
+            }
+            else
+            {
+                logger.LogWarning(RoslynResources.EAI_PluginResourceNotFound, plugin.Key, plugin.Version, plugin.StaticResourceName);
+            }
+        }
+
+        /// <summary>
+        /// Returns a version-specific directory in which the plugin resources
+        /// will be cached
+        /// </summary>
+        public static string GetPluginSpecificDir(string baseDir, Plugin plugin)
+        {
+            // Format is: [base]\[plugin.version]
+            string pluginVersionFolder = string.Format(System.Globalization.CultureInfo.InvariantCulture,
+                "{0}.{1}",
+                plugin.Key, plugin.Version);
+
+            pluginVersionFolder = StripInvalidDirectoryChars(pluginVersionFolder);
+
+            string fullDir = Path.Combine(baseDir, pluginVersionFolder);
+            return fullDir;
+        }
+
+        /// <summary>
+        /// Returns a unique directory for the specific resource
+        /// </summary>
+        public static string GetResourceSpecificDir(string baseDir, Plugin plugin)
+        {
+            // Format is: [base]\[plugin.version]\[resource name]
+
+            string pluginDir = GetPluginSpecificDir(baseDir, plugin);
+            string resourceFolder = StripInvalidDirectoryChars(plugin.StaticResourceName);
+
+            string fullDir = Path.Combine(pluginDir, resourceFolder);
+            return fullDir;
+        }
+
+
+        private static string StripInvalidDirectoryChars(string folderName)
+        {
+            return StripInvalidChars(folderName, Path.GetInvalidPathChars());
+        }
+
+        private static string StripInvalidFileChars(string folderName)
+        {
+            return StripInvalidChars(folderName, Path.GetInvalidFileNameChars());
+        }
```

Comment:
The unit tests don't currently cover invalid file or directory names. This could only happen the invalid character is in the NuGet package name or id. I'll add some tests.